src/pool/pool_disjoint.c

```diff
@@ -0,0 +1,1185 @@
+    bool *chunk = slab->chunks + slab->first_free_chunk_idx;
+    while (chunk != slab->chunks + slab->num_chunks) {
+        // false means not used
+        if (*chunk == false) {
+            size_t idx = (chunk - slab->chunks) / sizeof(bool);
+            LOG_DEBUG("idx: %zu\n", idx);
+            return idx;
+        }
+        chunk++;
+    }
+
+    LOG_DEBUG("idx: SIZE_MAX\n");
+    return SIZE_MAX;
+}
+
+void *slab_get_chunk(slab_t *slab) {
+    // free chunk must exist, otherwise we would have allocated another slab
+    const size_t chunk_idx = slab_find_first_available_chunk_idx(slab);
+    assert(chunk_idx != SIZE_MAX);
+
+    void *free_chunk =
+        (uint8_t *)slab->mem_ptr + chunk_idx * slab->bucket->size;
+
+    // mark chunk as used
+    slab->chunks[chunk_idx] = true;
+    slab->num_allocated += 1;
+
+    // use the found index as the next hint
+    slab->first_free_chunk_idx = chunk_idx + 1;
+
+    return free_chunk;
+}
+
+void *slab_get(const slab_t *slab) { return slab->mem_ptr; }
+void *slab_get_end(const slab_t *slab) {
+    return (uint8_t *)slab->mem_ptr + bucket_slab_min_size(slab->bucket);
+}
+
+void slab_free_chunk(slab_t *slab, void *ptr) {
+    // This method should be called through bucket (since we might remove the
+    // slab as a result), therefore all locks are done on that level.
+
+    // Make sure that we're in the right slab
+    assert(ptr >= slab_get(slab) && ptr < slab_get_end(slab));
+
+    // Even if the pointer p was previously aligned, it's still inside the
+    // corresponding chunk, so we get the correct index here.
+    size_t chunk_idx =
+        ((uint8_t *)ptr - (uint8_t *)slab->mem_ptr) / slab->bucket->size;
+
+    // Make sure that the chunk was allocated
+    assert(slab->chunks[chunk_idx] && "double free detected");
+    slab->chunks[chunk_idx] = false;
+    slab->num_allocated -= 1;
+
+    if (chunk_idx < slab->first_free_chunk_idx) {
+        slab->first_free_chunk_idx = chunk_idx;
+    }
+
+    LOG_DEBUG("chunk_idx: %zu, num_allocated: %zu, "
+              "first_free_chunk_idx: %zu\n",
+              chunk_idx, slab->num_allocated, slab->first_free_chunk_idx);
+}
+
+bool slab_has_avail(const slab_t *slab) {
+    return slab->num_allocated != slab->num_chunks;
+}
+
+void slab_reg(slab_t *slab) {
+    bucket_t *bucket = slab->bucket;
+
+    // NOTE: changed vs original DisjointPool implementation - currently slab
+    // is already aligned to bucket_slab_min_size. Additionally the end addr
+    // points to the last byte of slab data
+    void *start_addr = (void *)ALIGN_DOWN((size_t)slab_get(slab),
+                                          bucket_slab_min_size(bucket));
+    void *end_addr = (uint8_t *)(start_addr) + bucket_slab_min_size(bucket) - 1;
+
+    LOG_DEBUG("slab: %p, start: %p, end %p\n", (void *)slab, start_addr,
+              end_addr);
+
+    slab_reg_by_addr(start_addr, slab);
+    slab_reg_by_addr(end_addr, slab);
```

Comment:
why are we calling slab_reg_by_addr on end_addr?

Reply:
removed